File: VideoReadWriteDisplay/videoWrite.cpp
#include <opencv2/highgui.hpp>
#include <opencv2/videoio.hpp>

#include <cmath>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <string>

int main(int argc, char** argv) {
    const std::string keys =
        "{help h usage ?||Show this help message}"
        "{input|chaplin.mp4|Input video path}"
        "{output|output.avi|Output video path}"
        "{codec|MJPG|Four-character codec}"
        "{fps|-1|Output FPS; -1 preserves source FPS}"
        "{display||Show frames while writing}"
        "{max-frames|-1|Maximum frames; -1 writes all}";
    cv::CommandLineParser parser(argc, argv, keys);
    if (parser.has("help")) {
        parser.printMessage();
        return 0;
    }
    if (!parser.check()) {
        parser.printErrors();
        return 2;
    }

    try {
        const std::string inputPath = parser.get<std::string>("input");
        const std::string outputPath = parser.get<std::string>("output");
        const std::string codec = parser.get<std::string>("codec");
        const double requestedFps = parser.get<double>("fps");
        const int maxFrames = parser.get<int>("max-frames");
        const bool display = parser.has("display");
        if (codec.size() != 4) {
            throw std::invalid_argument(
                "codec must contain exactly four characters.");
        }
        if (!std::isfinite(requestedFps) ||
            (requestedFps != -1.0 && requestedFps <= 0.0)) {
            throw std::invalid_argument("fps must be positive or -1.");
        }
        if (maxFrames == 0 || maxFrames < -1) {
            throw std::invalid_argument("max-frames must be positive or -1.");
        }

        cv::VideoCapture capture(inputPath);
        if (!capture.isOpened()) {
            throw std::runtime_error("Could not open input video: " + inputPath);
        }
        cv::Mat frame;
        if (!capture.read(frame) || frame.empty()) {
            throw std::runtime_error("No decodable frames found.");
        }
        const cv::Size frameSize = frame.size();

        double sourceFps = capture.get(cv::CAP_PROP_FPS);
        if (!(sourceFps > 0.0)) {
            sourceFps = 30.0;
        }
        const double outputFps =
            requestedFps > 0.0 ? requestedFps : sourceFps;
        const int fourcc = cv::VideoWriter::fourcc(
            codec[0], codec[1], codec[2], codec[3]);
        cv::VideoWriter writer(
            outputPath, fourcc, outputFps, frameSize);
        if (!writer.isOpened()) {
            throw std::runtime_error(
                "Could not open output video: " + outputPath);
        }

        int framesWritten = 0;
        while (maxFrames < 0 || framesWritten < maxFrames) {
            writer.write(frame);
            ++framesWritten;
            if (display) {
                cv::imshow("Transcoding", frame);
                const int key = cv::waitKey(1) & 0xff;
                if (key == 27 || key == 'q') {
                    break;
                }
            }
            if (!capture.read(frame) || frame.empty()) {
                break;
            }
        }
        writer.release();

        std::cout << std::fixed << std::setprecision(4)
                  << "{\"frames_written\":" << framesWritten
                  << ",\"frame_size\":[" << frameSize.width << ','
                  << frameSize.height
                  << "],\"source_fps\":" << sourceFps
                  << ",\"output_fps\":" << outputFps
                  << ",\"codec\":\"" << codec
                  << "\",\"output\":\"" << outputPath << "\"}\n";
        return 0;
    } catch (const cv::Exception& error) {
        std::cerr << "OpenCV error: " << error.what() << '\n';
    } catch (const std::exception& error) {
        std::cerr << "error: " << error.what() << '\n';
    }
    return 2;
}

File: VideoReadWriteDisplay/videoWrite.py
#!/usr/bin/env python3
"""Transcode a video with checked OpenCV VideoCapture and VideoWriter objects."""

from __future__ import annotations

import argparse
import json
import math
from pathlib import Path
from typing import Any

import cv2


def transcode_video(
    input_path: Path,
    output_path: Path,
    *,
    codec: str = "MJPG",
    output_fps: float | None = None,
    display: bool = False,
    max_frames: int | None = None,
) -> dict[str, Any]:
    if len(codec) != 4:
        raise ValueError("codec must contain exactly four characters")
    if output_fps is not None and (
        not math.isfinite(output_fps) or output_fps <= 0
    ):
        raise ValueError("output_fps must be positive")
    if max_frames is not None and max_frames <= 0:
        raise ValueError("max_frames must be positive")

    capture = cv2.VideoCapture(str(input_path))
    if not capture.isOpened():
        raise FileNotFoundError(f"Could not open input video: {input_path}")

    writer: cv2.VideoWriter | None = None
    try:
        ok, first_frame = capture.read()
        if not ok or first_frame is None:
            raise RuntimeError(f"No decodable frames found in: {input_path}")

        height, width = first_frame.shape[:2]
        source_fps = capture.get(cv2.CAP_PROP_FPS)
        fps = output_fps or (source_fps if source_fps > 0 else 30.0)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        writer = cv2.VideoWriter(
            str(output_path),
            cv2.VideoWriter_fourcc(*codec),
            fps,
            (width, height),
        )
        if not writer.isOpened():
            raise RuntimeError(
                f"Could not open output video '{output_path}' with codec {codec}"
            )

        frames_written = 0
        frame = first_frame
        while max_frames is None or frames_written < max_frames:
            writer.write(frame)
            frames_written += 1
            if display:
                cv2.imshow("Transcoding", frame)
                if cv2.waitKey(1) & 0xFF in (27, ord("q")):
                    break
            ok, frame = capture.read()
            if not ok or frame is None:
                break

        return {
            "frames_written": frames_written,
            "frame_size": [width, height],
            "source_fps": source_fps,
            "output_fps": fps,
            "codec": codec,
            "output": str(output_path),
        }
    finally:
        capture.release()
        if writer is not None:
            writer.release()
        if display:
            cv2.destroyAllWindows()


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--input",
        type=Path,
        default=Path(__file__).parent / "chaplin.mp4",
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=Path("output.avi"),
    )
    parser.add_argument("--codec", default="MJPG")
    parser.add_argument("--fps", type=float)
    parser.add_argument("--max-frames", type=int)
    parser.add_argument("--display", action="store_true")
    return parser


def main() -> int:
    args = build_parser().parse_args()
    try:
        summary = transcode_video(
            args.input,
            args.output,
            codec=args.codec,
            output_fps=args.fps,
            display=args.display,
            max_frames=args.max_frames,
        )
    except (FileNotFoundError, RuntimeError, ValueError) as exc:
        print(f"error: {exc}")
        return 2
    print(json.dumps(summary, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
